tests/naive.py

```diff
@@ -186,40 +199,44 @@ def stump(T_A, m, T_B=None, exclusion_zone=None):
     else:
         diags = np.arange(-(n_A - m + 1) + 1, n_B - m + 1)
 
-    P = np.full((l, 3), np.inf)
-    I = np.full((l, 3), -1, dtype=np.int64)
+    P = np.full((l, k + 2), np.inf)
+    I = np.full((l, k + 2), -1, dtype=np.int64)  # two more columns are to store
+    # ... left and right top-1 matrix profile indices.
 
-    for k in diags:
-        if k >= 0:
-            iter_range = range(0, min(n_A - m + 1, n_B - m + 1 - k))
+    for g in diags:
+        if g >= 0:
+            iter_range = range(0, min(n_A - m + 1, n_B - m + 1 - g))
         else:
-            iter_range = range(-k, min(n_A - m + 1, n_B - m + 1 - k))
+            iter_range = range(-g, min(n_A - m + 1, n_B - m + 1 - g))
 
         for i in iter_range:
-            D = distance_matrix[i, i + k]
-            if D < P[i, 0]:
-                P[i, 0] = D
-                I[i, 0] = i + k
+            D = distance_matrix[i, i + g]
+            if D < P[i, k - 1]:
+                idx = searchsorted(P[i, :k], D)
+                # to keep the top-k, we need to the get rid of the last element.
+                P[i, :k] = np.insert(P[i, :k], idx, D)[:-1]
+                I[i, :k] = np.insert(I[i, :k], idx, i + g)[:-1]
 
             if ignore_trivial:  # Self-joins only
-                if D < P[i + k, 0]:
-                    P[i + k, 0] = D
-                    I[i + k, 0] = i
+                if D < P[i + g, k - 1]:
+                    idx = searchsorted(P[i + g, :k], D)
+                    P[i + g, :k] = np.insert(P[i + g, :k], idx, D)[:-1]
+                    I[i + g, :k] = np.insert(I[i + g, :k], idx, i)[:-1]
 
-                if i < i + k:
+                if i < i + g:
                     # Left matrix profile and left matrix profile index
-                    if D < P[i + k, 1]:
-                        P[i + k, 1] = D
-                        I[i + k, 1] = i
+                    if D < P[i + g, k]:
+                        P[i + g, k] = D
+                        I[i + g, k] = i
 
-                    if D < P[i, 2]:
+                    if D < P[i, k + 1]:
                         # right matrix profile and right matrix profile index
-                        P[i, 2] = D
-                        I[i, 2] = i + k
+                        P[i, k + 1] = D
+                        I[i, k + 1] = i + g
 
-    result = np.empty((l, 4), dtype=object)
-    result[:, 0] = P[:, 0]
-    result[:, 1:4] = I[:, :]
+    result = np.empty((l, 2 * k + 2), dtype=object)
+    result[:, :k] = P[:, :k]
```

Comment:
I think we had discussed that we should return:

```
result[:, 0] = P[:, 0]
result[:, 1] = I[:, 0]
result[:, 2:4] = I[:, -2:]
if k > 1:
    result[:, 4:4+k] = P[1:]
    result[:, 4+k:] = I[1:-2]
```

Reply:
Yeah...but, then I got to the similar code snippet you provided above and I thought it is hard to digest. 

**Then, I thought:** 
"I am adding `k=1` as the LAST argument in the function so I thought things should be fine. If `k=1`, then the output is similar to the existing `stumpy.stump`. If `k>1`, it can mean that the user probably understand what `k` is. If the user is not looking for left and right indices, they should be fine by simply getting the `first 2k columns`  which is super easy. IF they look for left and right indices, they can again easily get access to them by using negative index `-1` and `-2`."

What do you think about my justification? 

Btw, I understand it is probably a personal preference on how much a program should handle / take into account users' mistakes. If you do not agree with my justification, I will apply your proposed changes in next set of commits.

Reply:
So, I agree that it is more intuitive of a data structure to have everything next to each other. And I can accept that we can put a clear note in the docstring for stumpy.stump to clarify what the output is when k>1. I am wondering if we want to also return top-k left and right indices? So the result is simply 4*k 

Thoughts?

Reply:
Very interesting! In fact, I thought about returning top-k left and right indices as well. But, then, I thought the output can get really huge when I took a look at VALMOD paper again. So, for instance, if we return top-50 for a time series with length 1 million (as suggested in paper VALMOD), our output will have about `4 * 50 * 1e06` elements! 

I get your point.  the first-four-columns structure may help with easier development in the future. But, then, in that case, we can say the left indices are just `output[:, 2 * k : 3 * k]` (.the third `k` columns) which seems easy (maybe it is just easy in my head.)

Reply:
Yeah, I agree with the size being too big too. Okay, let's NOT do top-K left/right indices and let's keep what you're proposing in your code.

Reply:
> Furthermore the output of k-NN search contains a lot of trivial matches.
Assume a timeseries with only 1 motif-pair (A,B):
(A, B) are index locations
Then output of kNN (A) will be : [B, B-1, B+1, B-2, B+2, ....]
Due to the high number of trivial matches, the number k must also accomodate the wasted space, occupied by trivial neighbors.

Oh wow! I hadn't even considered that. That's pretty nasty :(

Reply:
> You can compress the output of k-NN search heavily.
uint32 is sufficently large for knn locations.
2^32 is 4.9E09, wheras the bechmark on the startpage shows up to lenght=1.4E08
float32 is sufficient for the knn distances, because even if the quantization would result in distance-ties, then the correct order of neighbors is also encoded in the position of the output, because the output is sorted by distance.

Would you elaborate this?

> Furthermore the output of k-NN search contains a lot of trivial matches.
Assume a timeseries with only 1 motif-pair (A,B):
(A, B) are index locations
Then output of kNN (A) will be : [B, B-1, B+1, B-2, B+2, ....]
Due to the high number of trivial matches, the number k must also accomodate the wasted space, occupied by trivial neighbors.

Yes, that is correct! I THINK there are probably two approaches for k-NN matrix profile. Please allow me elaborate.

* K-NN ignoring trivial matches (similar to finding top-k matches of query while ignoring trivial matches)
* K-NN without ignoring trivial matches.

As you noted, this PR is the latter one. However, this is not necessarily bad. The ultimate goal is to use this meta info (extracted for subsequences with length `m`) to find matrix profile of length `m+1`. 

Let's say, for length `m` , we have `KNN(A) = B `. There is a high chance that for length `m+1`, KNN(A) might be in [B-2, B-1, B, B+1, B+2]. I am not saying that K-NN always return trivial zones. What I am trying to say is that such output is not necessarily bad for what it tries to achieve: "accelerating the calculation of matrix profile of subsequences with `length > m` "

So, if someone is looking for matrix profile of length `m` only.... things should be fine and they can return top-1 NN. However, IF a user is looking for EXACT matrix profile for a range of subsequence length, then he will probably get benefit out of it. 

As you noticed, there are some trivial matches. So, if a user wants to do some research on discovering motifs/discords of one length only, and, let's say, they want to see what happens if they consider group of patterns, something similar to twin-freak phenomenon, then, they should ignore trivial matches for that kind of research.